src/apple/macos/component/arm.rs

```diff
@@ -0,0 +1,149 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use std::ffi::CStr;
+
+use core_foundation_sys::array::CFArrayGetCount;
+use core_foundation_sys::array::CFArrayGetValueAtIndex;
+use core_foundation_sys::base::kCFAllocatorDefault;
+use core_foundation_sys::string::kCFStringEncodingUTF8;
+use core_foundation_sys::string::CFStringCreateWithBytes;
+use core_foundation_sys::string::CFStringGetCStringPtr;
+
+use crate::apple::inner::ffi::kHIDPage_AppleVendor;
+use crate::apple::inner::ffi::kHIDUsage_AppleVendor_TemperatureSensor;
+use crate::apple::inner::ffi::kIOHIDEventTypeTemperature;
+use crate::apple::inner::ffi::matching;
+use crate::apple::inner::ffi::IOHIDEventFieldBase;
+use crate::apple::inner::ffi::IOHIDEventGetFloatValue;
+use crate::apple::inner::ffi::IOHIDEventSystemClientCopyServices;
+use crate::apple::inner::ffi::IOHIDEventSystemClientCreate;
+use crate::apple::inner::ffi::IOHIDEventSystemClientSetMatching;
+use crate::apple::inner::ffi::IOHIDServiceClientCopyEvent;
+use crate::apple::inner::ffi::IOHIDServiceClientCopyProperty;
+use crate::apple::inner::ffi::IOHIDServiceClientRef;
+use crate::apple::inner::ffi::HID_DEVICE_PROPERTY_PRODUCT;
+use crate::ComponentExt;
+
+pub(crate) fn temperatures(components: &mut Vec<Component>) -> &mut Vec<Component> {
+    components.clear();
+
+    unsafe {
+        let matches = matching(
+            kHIDPage_AppleVendor,
+            kHIDUsage_AppleVendor_TemperatureSensor,
+        );
+        if matches.is_null() {
+            return components;
+        }
+
+        let client = IOHIDEventSystemClientCreate(kCFAllocatorDefault);
+        if client.is_null() {
+            return components;
+        }
+
+        let _ = IOHIDEventSystemClientSetMatching(client, matches);
+
+        let services = IOHIDEventSystemClientCopyServices(client);
+        if services.is_null() {
+            return components;
+        }
+
+        let key_ref = CFStringCreateWithBytes(
```

Comment:
A lot of `free` are missing. For example `key_ref` should be freed using `CFRelease` once you're done converting it into Rust.

Reply:
I am not sure which and when to free the pointer got from the core foundation lib and I did some tests and rewrite this part, but I am still not sure I did it right

Reply:
Yes, by using the `Drop` trait is a better way to free the resource. 
The problem is I am not familiar with Core Foundation API, it seems some pointer return from it does not need to be freed by the caller. 
After some search I found this page [Ownership Policy](https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFMemoryMgmt/Concepts/Ownership.html#//apple_ref/doc/uid/20001148-CJBEJBHH), and it is very helpful

Reply:
Ah nice! Don't hesitate to put the link on the `PtrWrap` if you use it.